apps/system/js/launcher.js

```diff
@@ -0,0 +1,108 @@
+  Launcher.SUB_MODULES = ['Scheduler'];
+  Launcher.SETTINGS = [
+    'ftu.manifestURL',
+    'homescreen.manifestURL',
+    'deviceinfo.previous_os',
+    'deviceinfo.os',
+    'lockscreen.enabled'
+  ];
+  Launcher.STATES = [
+    'justUpgraded'
+  ];
+  BaseModule.create(Launcher, {
+    name: 'Launcher',
+    DEBUG: false,
+    _start: function() {
+      this._version = {};
+      this.debug('reading configurations...');
+      return Promise.all([
+        this.readFtu(),
+        this.readSetting('homescreen.manifestURL'),
+        this.readSetting('ftu.manifestURL'),
+        this.readSetting('deviceinfo.os'),
+        this.readSetting('deviceinfo.previous_os'),
+        this.readSetting('lockscreen.enabled')
+      ]).then(function(results) {
+        var [ftuEnabled, homescreenManifestURL, ftuManifestURL,
+             osVersion, previousOsVersion, lockscreenEnabled] = results;
+        this.lockscreenEnabled = lockscreenEnabled;
+        this.debug('Ftu enabled = ' + ftuEnabled +
+          '; ftuManifestURL = ' + ftuManifestURL +
+          '; osVersion = ' + osVersion +
+          '; previousOsVersion = ' + previousOsVersion +
+          '; homescreenManifestURL = ' + homescreenManifestURL +
+          '; lockscreenEnabled = ' + lockscreenEnabled);
+        if (this.checkUpgrading(osVersion, previousOsVersion)) {
+          this.debug('upgrading boot');
+          this.launchFtuThenHomescreen(ftuManifestURL, homescreenManifestURL);
+        } else {
+          if (ftuEnabled !== false && ftuManifestURL) {
+            this.debug('clean boot');
+            this.launchFtuThenHomescreen(ftuManifestURL, homescreenManifestURL);
+          } else {
+            this.debug('normal boot');
+            // We still need to tell FtuLauncher to skip to process some tasks.
+            this.service.request('FtuLauncher:skip');
+            if (lockscreenEnabled) {
+              // XXX: We should have a lockscreenLauncher here.
+              this.service.request('LockScreenLauncher:launch').then(() => {
+                this.service.request('LogoManager:animatePoweronLogo');
+                this.service.request('HomescreenLauncher:launch',
+                  homescreenManifestURL);
+              });
+            } else {
+              this.service.request('HomescreenLauncher:launch',
+                homescreenManifestURL).then(() => {
+                  this.service.request('LogoManager:animatePoweronLogo');
+                });
+              this.service.request('LockScreenLauncher:standby');
+            }
+          }
+        }
+      }.bind(this));
+    },
+    justUpgraded: function() {
+      return this._isUpgrade;
+    },
+    checkUpgrading: function(prev, curr) {
+      var isUpgrade = false;
+      // dont treat lack of previous version info as an upgrade
+      if (prev && curr) {
+        isUpgrade = curr.major > prev.major || curr.minor > prev.minor;
+      }
+      this._isUpgrade = isUpgrade;
```

Comment:
Well, even though I know your design is based on mixing, to see `this` in a literal object still make me worrisome. It's like those old "Singleton" that expect its `this` would correctly bind to the object, while actually they don't do anything to ensure that excepts to add event listener with the object as the context, which is very implicit and sometimes (especially in tests) they get broken.

From BaseModule it looks compatible with the current prototype-constructor pattern. I don't know if you think it's good to keep the pattern with the new BaseModule, but since we have had such pattern in our codebase, IMHO it's better to design things follow the current style, rather to mix these two patterns in our codebase.